### PR TITLE
Output shape and dtype inference, multioutput inner layers

### LINKED ISSUE #13: tensorflow 2.3.1 is not supported
The change introduced in PR #12 calls
```python
from tensorflow.python.util.keras_deps import get_call_context_function
```

The module `keras_deps` not exist in tensorflow 2.3.1 (which should be supported according to the requirements).

### PR DESCRIPTION
**Pull request by DavideWalder**
### Changes
- Fixed output shape when running in Keras functional mode. The ragged dimension was missing returning (batch, features...) instead of (batch, ragged dim, features...)
- Expanded test suite to Python 3.9 and Tensorflow 2.5
- Dropped support for Tensorflow<2.4 (closes #13)
- Removed `_type_spec`
- Added `compute_output_shape` (used internally and by other layers to infer the output shape without using `call`)
- Added support for multi-output inner layers.
- Fixed dtype: now multiple output dtypes other than int64 are supported, correctly inferred and predicted.
- Fixed issue where the `ListMapper` wouldn't work if the input's row_split dtype wasn't `tf.int64`

### Notes
- When running `compute_output_shape` on an inner layer that doesn't override it, Keras will run its `call` function with a non-ragged placeholder. This broke the assertions in `FakeLayer`, which was removed. Alternatively, a `compute_output_shape` method could have been defined.
- The implementation for the new features was kept as close as possible to `TimeDistributed`, for consistency.
- RaggedTensorSpec has properties such as `row_splits_dtype` and `ragged_rank` which are not represented in the placeholder returned by `ListMapper.__call__`.

### CODE REVIEW

File: keras_list_mapper/keras_list_mapper.py [hunk 1]
```diff
@@ -111,73 +114,75 @@ def __call__(self, inputs, *args, **kwargs):
...
             self._built_from_signature = True
 
-        super().build(inputs)
+        # Build layer and inner layer
+        self.build(input_shape)
+        output_shapes = self.compute_output_shape(input_shape)
+
+        def replace_shape(tensor, shape: tf.TensorShape):
+            return K.placeholder(shape, dtype=tensor.dtype, ragged=True)
+        outputs = nest.map_structure(replace_shape, outputs, output_shapes)
```
Comment (pedro-r-marques) [inline keras_list_mapper/keras_list_mapper.py]: Is this required ? My concern is that the tensor outputs from calling the inner layer are discarded.
While working on this previously keras was rather keen on having the outputs of the layer connected to the next layer. That is the reason _set_connectivity_metadata is called... without it the graph is considered to be disconnected. I don't think the outputs should be overriden.

I understand that we want to make the outputs a ragged tensor but it would be nice to find a way to do that without using a placeholder.

File: keras_list_mapper/keras_list_mapper.py [hunk 2]
```diff
@@ -211,16 +217,65 @@ def call(self, inputs, *args, **kwargs):  # pylint: disable=unused-argument
...
+            child_output_shape = child_output_shape[1:]
+            if len(child_output_shape) == 1:
+                child_output_shape = child_output_shape[0]
+        child_output_shape = tf_utils.convert_shapes(
+            child_output_shape, to_tuples=False)
+        timesteps = nest.flatten(tf_utils.convert_shapes(input_shape))[1]
+
+        def insert_timesteps(dims: tf.TensorShape):
+            dims = dims.as_list()
+            return tf.TensorShape([dims[0], timesteps] + dims[1:])
```
Comment (pedro-r-marques) [inline keras_list_mapper/keras_list_mapper.py:251]: The ListMapper returns a ragged Tensor. There is a new layer API compute_output_signature that is suppose to be able to convey this information. Would it makes sense to use that instead ?

File: keras_list_mapper/keras_list_mapper.py (hunk 1 above)
Comment (DavideWalder) [reply keras_list_mapper/keras_list_mapper.py]: My comment on `_set_connectivity_metadata` is wrong, it is indeed responsible for connecting the graph.

Regarding the connection with the inner layer: it seems that wrappers are not connected to their inner layer. In order for a graph to be properly connected, inputs and outputs of each layer must be connected. Since the inner layer is not part of the graph this doesn't seem to be a problem.

Example:
```python
inputs = layers.Input((None, 1), ragged=True)
td = layers.TimeDistributed(layers.Dense(1))(inputs)
lstm = layers.LSTM(1)(td)
model = models.Model(inputs, lstm)
assert len(model.layers[-2]._outbound_nodes) == 1
assert len(model.layers[-2].layer._outbound_nodes) == 0
```
```python
inputs = layers.Input((None, 1), ragged=True)
lm = ListMapper(layers.Dense(1))(inputs)
lstm = layers.LSTM(1)(lm)
model = models.Model(inputs, lstm)
assert len(model.layers[-2]._outbound_nodes) == 1
assert len(model.layers[-2].layer._outbound_nodes) == 0
```

The best solution of course would be not to override `__call__` and use `call` also when building the graph. But with the current implementation this is still not possible. [Dynamic layers](https://github.com/tensorflow/tensorflow/blob/ec5ff5df834bcd32941ffac649950558f2adea1f/tensorflow/python/keras/engine/base_layer.py#L864) are built in a similar fashion.

File: keras_list_mapper/keras_list_mapper.py (hunk 2 above)
Comment (DavideWalder) [reply keras_list_mapper/keras_list_mapper.py:251]: Even though `compute_output_signature` seems to be called only for [dynamic layers](https://github.com/tensorflow/tensorflow/blob/ec5ff5df834bcd32941ffac649950558f2adea1f/tensorflow/python/keras/engine/base_layer.py#L864), it is indeed useful internally.
It is now implemented and used to create the output placeholders, simplifying the code in `__call__`.

**Review by pedro-r-marques — APPROVED** [review]

File: keras_list_mapper/keras_list_mapper.py (hunk 2 above)
Comment (DavideWalder) [reply keras_list_mapper/keras_list_mapper.py:251]: Even though `compute_output_signature` seems to be called only for [dynamic layers](https://github.com/tensorflow/tensorflow/blob/ec5ff5df834bcd32941ffac649950558f2adea1f/tensorflow/python/keras/engine/base_layer.py#L864), it is indeed useful internally.
It is now implemented and used to create the output placeholders, simplifying the code in `__call__`.
